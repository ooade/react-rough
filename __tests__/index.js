import React from 'react';
import { configure, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import ReactRough, {
	NodeMounter,
	Arc,
	Circle,
	Curve,
	Ellipse,
	Line,
	LinearPath,
	Path,
	Polygon,
	Rectangle
} from '../src';

configure({ adapter: new Adapter() });

describe('ReactRough', () => {
	let consoleError;

	beforeAll(() => {
		jest.spyOn(window._virtualConsole, 'emit').mockImplementation(() => false);

		consoleError = jest.spyOn(console, 'error').mockImplementation(() => false);
	});

	afterAll(() => {
		consoleError.mockClear();
	});

	describe('Core', () => {
		it('should render properly', () => {
			const wrapper = render(<ReactRough width={200} height={400} />);
			expect(wrapper).toMatchSnapshot();
		});

		it('should render properly with svg renderer', () => {
			const wrapper = render(<ReactRough width={200} height={400} renderer="svg" />);
			expect(wrapper).toMatchSnapshot();
		});

		it('should clear the canvas on first render', () => {
			const spy = jest.spyOn(ReactRough.prototype, 'clearCanvas');
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Circle points={[50, 50, 80]} fill="red" />
				</ReactRough>
			);
			expect(wrapper.instance().ctx).not.toBeNull()
			expect(spy).toHaveBeenCalledTimes(2);
			spy.mockRestore();
		});

		it('should clear the canvas on redraw', () => {
			const spy = jest.spyOn(ReactRough.prototype, 'clearCanvas');
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Circle points={[50, 50, 80]} fill="red" />
				</ReactRough>
			);
			expect(spy).toHaveBeenCalledTimes(2);
			wrapper.instance().redraw();
			expect(spy).toHaveBeenCalledTimes(3);
			spy.mockRestore();
		});

		it('should use backgroundColor as canvas background color', () => {
			const wrapper = mount(
				<ReactRough
					width={200}
					height={400}
					backgroundColor="rgba(126, 255, 0, 0.1)"
				>
					<Circle points={[50, 50, 80]} fill="red" />
				</ReactRough>
			);
			expect(wrapper).toMatchSnapshot();
		});

		it('should use backgroundColor as svg background color', () => {
			const wrapper = mount(
				<ReactRough
					width={200}
					height={400}
					backgroundColor="rgba(126, 255, 0, 0.1)"
					renderer="svg"
				/>
			);
			expect(wrapper).toMatchSnapshot();
		});

		it('should render properly with children', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Circle points={[50, 50, 80]} fill="red" />
				</ReactRough>
			);
			expect(wrapper).toMatchSnapshot();
		});

		it('should render properly when nested within other components', () => {
			const SomeComponent = () => <div>Some Component </div>;

			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<SomeComponent>
						<Circle points={[50, 50, 80]} fill="red" />
					</SomeComponent>
				</ReactRough>
			);
			expect(wrapper).toMatchSnapshot();
		});

		it('should throw an error if the Provider component isnt found', () => {
			expect(() =>
				mount(<Circle points={[50, 50, 80]} fill="red" />)
			).toThrowError('ReactRough Component not found!');
		});

		it('should throw an error if an invalid key is provided to a static component', () => {
			expect(() =>
				mount(
					<ReactRough width={200} height={400}>
						<Circle points={[50, 50, 80]} fsill="red" />
					</ReactRough>
				)
			).toThrowError('Invalid key "fsill" assigned to "circle component"');
		});

		it('should update svg as the props update', () => {
			const cDMspy = jest.spyOn(NodeMounter.prototype, 'componentDidMount');
			const cDUspy = jest.spyOn(NodeMounter.prototype, 'componentDidUpdate');
			const cWUspy = jest.spyOn(NodeMounter.prototype, 'componentWillUnmount');
			const wrapper = mount(
				<ReactRough width={200} height={400} renderer="svg">
					<Circle points={[50, 50, 80]} fill="red" />
				</ReactRough>
			);
			expect(cDMspy).toHaveBeenCalledTimes(1)
			expect(cDUspy).toHaveBeenCalledTimes(0)
			expect(cWUspy).toHaveBeenCalledTimes(0)
			const firstRender = wrapper.find('g').html();

			wrapper.setProps({
				children: React.cloneElement(wrapper.props().children, { points: [60, 60, 90] })
			});
			expect(cDMspy).toHaveBeenCalledTimes(1)
			expect(cDUspy).toHaveBeenCalledTimes(1)
			expect(cWUspy).toHaveBeenCalledTimes(0)

			const secondRender = wrapper.find('g').html();
			expect(secondRender).not.toEqual(firstRender);

			wrapper.unmount();
			expect(cDMspy).toHaveBeenCalledTimes(1)
			expect(cDUspy).toHaveBeenCalledTimes(1)
			expect(cWUspy).toHaveBeenCalledTimes(1)
		});
	});

	describe('Arc', () => {
		it('should render properly with props', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Arc points={[50, 50, 80]} fill="red" />
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('Circle', () => {
		it('should render properly with props', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Circle points={[50, 50, 80]} fill="red" />
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('Curve', () => {
		it('should render properly with props', () => {
			// draw sine curve
			let points = [];
			for (let i = 0; i < 20; i++) {
				let x = (400 / 20) * i + 10;
				let xdeg = (Math.PI / 100) * x;
				let y = Math.round(Math.sin(xdeg) * 90) + 500;
				points.push([x, y]);
			}

			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Curve points={points} fill="red" />
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('Ellipse', () => {
		it('should render properly with props', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Ellipse points={[10, 50, 150, 80]} fill="blue" stroke="red" />
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('Line', () => {
		it('should render properly with props', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Line points={[60, 60, 190, 60]} fill="blue" stroke="red" />
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('Path', () => {
		it('should render properly with props', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Path
						dataString="M80 80 A 45 45, 0, 0, 0, 125 125 L 125 80 Z"
						fill="blue"
						stroke="red"
					/>
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});

		it('should throw error when points is used', () => {
			expect(() => {
				mount(
					<ReactRough width={200} height={400}>
						<Path
							points="M80 80 A 45 45, 0, 0, 0, 125 125 L 125 80 Z"
							fill="blue"
							stroke="red"
						/>
					</ReactRough>
				);
			}).toThrowError('You need a dataString property for path, not points');
		});
	});

	describe('Polygon', () => {
		it('should render properly with props', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Polygon
						points={[[690, 130], [790, 140], [750, 240], [690, 220]]}
						fill="blue"
						stroke="red"
					/>
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('Rectangle', () => {
		it('should render properly with props', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<Rectangle points={[10, 10, 100, 100]} fill="blue" stroke="red" />
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('LinearPath', () => {
		it('should render properly with props', () => {
			const wrapper = mount(
				<ReactRough width={200} height={400}>
					<LinearPath
						points={[[690, 130], [790, 140], [750, 240], [690, 220]]}
						fill="blue"
						stroke="red"
					/>
				</ReactRough>
			);

			expect(wrapper).toMatchSnapshot();
		});
	});
});
