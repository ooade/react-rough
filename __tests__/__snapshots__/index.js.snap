// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactRough Arc should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Arc
      fill="red"
      points={
        Array [
          50,
          50,
          80,
        ]
      }
    >
      <RoughConsumer
        fill="red"
        points={
          Array [
            50,
            50,
            80,
          ]
        }
        type="arc"
      />
    </Arc>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Circle should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Circle
      fill="red"
      points={
        Array [
          50,
          50,
          80,
        ]
      }
    >
      <RoughConsumer
        fill="red"
        points={
          Array [
            50,
            50,
            80,
          ]
        }
        type="circle"
      />
    </Circle>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Core should render properly 1`] = `
<canvas
  height="400"
  width="200"
/>
`;

exports[`ReactRough Core should render properly when nested within other components 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <SomeComponent>
      <div>
        Some Component 
      </div>
    </SomeComponent>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Core should render properly with children 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Circle
      fill="red"
      points={
        Array [
          50,
          50,
          80,
        ]
      }
    >
      <RoughConsumer
        fill="red"
        points={
          Array [
            50,
            50,
            80,
          ]
        }
        type="circle"
      />
    </Circle>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Core should render properly with svg backend 1`] = `
<svg
  height="400"
  width="200"
/>
`;

exports[`ReactRough Core should use backgroundColor as canvas background color 1`] = `
<ReactRough
  backend="canvas"
  backgroundColor="rgba(126, 255, 0, 0.1)"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Circle
      fill="red"
      points={
        Array [
          50,
          50,
          80,
        ]
      }
    >
      <RoughConsumer
        fill="red"
        points={
          Array [
            50,
            50,
            80,
          ]
        }
        type="circle"
      />
    </Circle>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Core should use backgroundColor as svg background color 1`] = `
<ReactRough
  backend="svg"
  backgroundColor="rgba(126, 255, 0, 0.1)"
  height={400}
  width={200}
>
  <svg
    height={400}
    style={
      Object {
        "backgroundColor": "rgba(126, 255, 0, 0.1)",
      }
    }
    width={200}
  />
</ReactRough>
`;

exports[`ReactRough Curve should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Curve
      fill="red"
      points={
        Array [
          Array [
            10,
            528,
          ],
          Array [
            30,
            573,
          ],
          Array [
            50,
            590,
          ],
          Array [
            70,
            573,
          ],
          Array [
            90,
            528,
          ],
          Array [
            110,
            472,
          ],
          Array [
            130,
            427,
          ],
          Array [
            150,
            410,
          ],
          Array [
            170,
            427,
          ],
          Array [
            190,
            472,
          ],
          Array [
            210,
            528,
          ],
          Array [
            230,
            573,
          ],
          Array [
            250,
            590,
          ],
          Array [
            270,
            573,
          ],
          Array [
            290,
            528,
          ],
          Array [
            310,
            472,
          ],
          Array [
            330,
            427,
          ],
          Array [
            350,
            410,
          ],
          Array [
            370,
            427,
          ],
          Array [
            390,
            472,
          ],
        ]
      }
    >
      <RoughConsumer
        fill="red"
        points={
          Array [
            Array [
              10,
              528,
            ],
            Array [
              30,
              573,
            ],
            Array [
              50,
              590,
            ],
            Array [
              70,
              573,
            ],
            Array [
              90,
              528,
            ],
            Array [
              110,
              472,
            ],
            Array [
              130,
              427,
            ],
            Array [
              150,
              410,
            ],
            Array [
              170,
              427,
            ],
            Array [
              190,
              472,
            ],
            Array [
              210,
              528,
            ],
            Array [
              230,
              573,
            ],
            Array [
              250,
              590,
            ],
            Array [
              270,
              573,
            ],
            Array [
              290,
              528,
            ],
            Array [
              310,
              472,
            ],
            Array [
              330,
              427,
            ],
            Array [
              350,
              410,
            ],
            Array [
              370,
              427,
            ],
            Array [
              390,
              472,
            ],
          ]
        }
        type="curve"
      />
    </Curve>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Ellipse should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Ellipse
      fill="blue"
      points={
        Array [
          10,
          50,
          150,
          80,
        ]
      }
      stroke="red"
    >
      <RoughConsumer
        fill="blue"
        points={
          Array [
            10,
            50,
            150,
            80,
          ]
        }
        stroke="red"
        type="ellipse"
      />
    </Ellipse>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Line should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Line
      fill="blue"
      points={
        Array [
          60,
          60,
          190,
          60,
        ]
      }
      stroke="red"
    >
      <RoughConsumer
        fill="blue"
        points={
          Array [
            60,
            60,
            190,
            60,
          ]
        }
        stroke="red"
        type="line"
      />
    </Line>
  </canvas>
</ReactRough>
`;

exports[`ReactRough LinearPath should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <LinearPath
      fill="blue"
      points={
        Array [
          Array [
            690,
            130,
          ],
          Array [
            790,
            140,
          ],
          Array [
            750,
            240,
          ],
          Array [
            690,
            220,
          ],
        ]
      }
      stroke="red"
    >
      <RoughConsumer
        fill="blue"
        points={
          Array [
            Array [
              690,
              130,
            ],
            Array [
              790,
              140,
            ],
            Array [
              750,
              240,
            ],
            Array [
              690,
              220,
            ],
          ]
        }
        stroke="red"
        type="linearPath"
      />
    </LinearPath>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Path should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Path
      dataString="M80 80 A 45 45, 0, 0, 0, 125 125 L 125 80 Z"
      fill="blue"
      stroke="red"
    >
      <RoughConsumer
        dataString="M80 80 A 45 45, 0, 0, 0, 125 125 L 125 80 Z"
        fill="blue"
        stroke="red"
        type="path"
      />
    </Path>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Polygon should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Polygon
      fill="blue"
      points={
        Array [
          Array [
            690,
            130,
          ],
          Array [
            790,
            140,
          ],
          Array [
            750,
            240,
          ],
          Array [
            690,
            220,
          ],
        ]
      }
      stroke="red"
    >
      <RoughConsumer
        fill="blue"
        points={
          Array [
            Array [
              690,
              130,
            ],
            Array [
              790,
              140,
            ],
            Array [
              750,
              240,
            ],
            Array [
              690,
              220,
            ],
          ]
        }
        stroke="red"
        type="polygon"
      />
    </Polygon>
  </canvas>
</ReactRough>
`;

exports[`ReactRough Rectangle should render properly with props 1`] = `
<ReactRough
  backend="canvas"
  height={400}
  width={200}
>
  <canvas
    height={400}
    width={200}
  >
    <Rectangle
      fill="blue"
      points={
        Array [
          10,
          10,
          100,
          100,
        ]
      }
      stroke="red"
    >
      <RoughConsumer
        fill="blue"
        points={
          Array [
            10,
            10,
            100,
            100,
          ]
        }
        stroke="red"
        type="rectangle"
      />
    </Rectangle>
  </canvas>
</ReactRough>
`;
